### Importing initial modules

In [1]:
import os
import numpy as np
import pandas as pd

In [2]:
os.getcwd()

'C:\\Users\\baari\\Desktop\\hard-drive-predictive-maintenance'

### Changing to the correct directory if not already

In [3]:
path = '/Users/baari/Desktop/hard-drive-predictive-maintenance'
os.chdir(path)
os.getcwd()

'C:\\Users\\baari\\Desktop\\hard-drive-predictive-maintenance'

In [22]:
df = pd.read_csv('Q1_2019.csv')

In [23]:
df.shape

(9577046, 129)

In [29]:
df.head()

date   serial_number                 model  capacity_bytes  failure  \
0  2019-01-01        Z305B2QN           ST4000DM000   4000787030016        0   
1  2019-01-01        ZJV0XJQ4         ST12000NM0007  12000138625024        0   
2  2019-01-01        ZJV0XJQ3         ST12000NM0007  12000138625024        0   
3  2019-01-01        ZJV0XJQ0         ST12000NM0007  12000138625024        0   
4  2019-01-01  PL1331LAHG1S4H  HGST HMS5C4040ALE640   4000787030016        0   

   smart_1_normalized  smart_1_raw  smart_2_normalized  smart_2_raw  \
0               111.0   35673128.0                 NaN          NaN   
1                83.0  187116872.0                 NaN          NaN   
2                73.0   19599104.0                 NaN          NaN   
3                81.0  136943696.0                 NaN          NaN   
4               100.0          0.0               134.0        103.0   

   smart_3_normalized  ...  smart_250_normalized  smart_250_raw  \
0                91.0  ...                   NaN            NaN   
1                98.0  ...                   NaN            NaN   
2                99.0  ...                   NaN            NaN   
3                93.0  ...                   NaN            NaN   
4               100.0  ...                   NaN            NaN   

   smart_251_normalized  smart_251_raw  smart_252_normalized  smart_252_raw  \
0                   NaN            NaN                   NaN            NaN   
1                   NaN            NaN                   NaN            NaN   
2                   NaN            NaN                   NaN            NaN   
3                   NaN            NaN                   NaN            NaN   
4                   NaN            NaN                   NaN            NaN   

   smart_254_normalized  smart_254_raw  smart_255_normalized  smart_255_raw  
0                   NaN            NaN                   NaN            NaN  
1                   NaN            NaN                   NaN            NaN  
2                   NaN            NaN                   NaN            NaN  
3                   NaN            NaN                   NaN            NaN  
4                   NaN            NaN                   NaN            NaN  

[5 rows x 129 columns]

In [30]:
df.tail()

date   serial_number                 model  capacity_bytes  \
9577041  2019-03-31  PL1331LAHD1AWH  HGST HMS5C4040BLE640   4000787030016   
9577042  2019-03-31        ZA10MCEQ           ST8000DM002   8001563222016   
9577043  2019-03-31        ZCH0CRTK         ST12000NM0007  12000138625024   
9577044  2019-03-31  PL1331LAHD1T5H  HGST HMS5C4040BLE640   4000787030016   
9577045  2019-03-31  PL2331LAHDS4TJ  HGST HMS5C4040BLE640   4000787030016   

         failure  smart_1_normalized  smart_1_raw  smart_2_normalized  \
9577041        0               100.0          0.0               134.0   
9577042        0                72.0   15233376.0                 NaN   
9577043        0                81.0  122099464.0                 NaN   
9577044        0               100.0          0.0               134.0   
9577045        0               100.0          0.0               133.0   

         smart_2_raw  smart_3_normalized  ...  smart_250_normalized  \
9577041        100.0               100.0  ...                   NaN   
9577042          NaN                94.0  ...                   NaN   
9577043          NaN                97.0  ...                   NaN   
9577044        101.0               148.0  ...                   NaN   
9577045        104.0               100.0  ...                   NaN   

         smart_250_raw  smart_251_normalized  smart_251_raw  \
9577041            NaN                   NaN            NaN   
9577042            NaN                   NaN            NaN   
9577043            NaN                   NaN            NaN   
9577044            NaN                   NaN            NaN   
9577045            NaN                   NaN            NaN   

         smart_252_normalized  smart_252_raw  smart_254_normalized  \
9577041                   NaN            NaN                   NaN   
9577042                   NaN            NaN                   NaN   
9577043                   NaN            NaN                   NaN   
9577044                   NaN            NaN                   NaN   
9577045                   NaN            NaN                   NaN   

         smart_254_raw  smart_255_normalized  smart_255_raw  
9577041            NaN                   NaN            NaN  
9577042            NaN                   NaN            NaN  
9577043            NaN                   NaN            NaN  
9577044            NaN                   NaN            NaN  
9577045            NaN                   NaN            NaN  

[5 rows x 129 columns]

In [24]:
df['failure'].value_counts()

failure
0    9576618
1        428
Name: count, dtype: int64

### Handling missing values

In [25]:
total_cells = np.prod(df.shape)
total_cells

1235438934

In [26]:
missing_values_count = df.isnull().sum()
missing_values_count[0:10]

date                        0
serial_number               0
model                       0
capacity_bytes              0
failure                     0
smart_1_normalized       1534
smart_1_raw              1534
smart_2_normalized    7457196
smart_2_raw           7457196
smart_3_normalized       3363
dtype: int64

In [27]:
total_missing = missing_values_count.sum()
total_missing

765053890

In [28]:
percent_missing = total_missing / total_cells * 100
print(f'{percent_missing}% of the data is missing')

61.925674264042584% of the data is missing


In [ ]:
columns_before = df.columns

# Remove columns that have all null values
df = df.dropna(axis=1, how='all')

columns_after = df.columns

removed_columns = set(columns_before) - set(columns_after)
print(f'Columns removed: {removed_columns}')

In [68]:
total_cells = np.prod(df.shape)
total_missing = df.isnull().sum().sum()
percent_missing = total_missing / total_cells * 100
print(f'{percent_missing}% of the data is missing')

57.290538957056455% of the data is missing


# Computing failure rates

### Creating a dataframe that has the number of drive days for each model. Drive days refers to the number of days a hard drive has been running (the number of rows in the main dataframe for that model).

### Written in SQL like this:

CREATE TABLE drive_days AS 
    SELECT model, count(*) AS drive_days 
    FROM drive_stats 
    GROUP BY model;

In [32]:
# Groups the dataframe by the 'model' column and calculates the size (number of rows) for each group
# The .size() gets the count of occurrences for each model
drive_days = df.groupby('model').size().reset_index(name='drive_days')

# Sort the dataframe by 'drive_days' in descending order
drive_days = drive_days.sort_values(by='drive_days', ascending=False).reset_index(drop=True)

drive_days

model  drive_days
0                         ST12000NM0007     2955025
1                           ST4000DM000     1989429
2                          ST8000NM0055     1294451
3                  HGST HMS5C4040BLE640     1172824
4                           ST8000DM002      888741
5                  HGST HMS5C4040ALE640      313383
6                  HGST HUH721212ALN604      259749
7                           ST6000DX000      135832
8                   TOSHIBA MG07ACA14TA      109404
9                         ST10000NM0086      108555
10                 HGST HUH728080ALE600       93598
11                        ST500LM012 HN       50619
12                   TOSHIBA MQ01ABF050       46969
13                  TOSHIBA MQ01ABF050M       32624
14                         WDC WD60EFRX       30523
15                       WDC WD5000LPVX       22015
16                           ST500LM030       14479
17                 HGST HUH721212ALE600       14040
18                  TOSHIBA MD04ABA400V       12662
19                       WDC WD5000LPCX        4920
20                          ST4000DM005        4848
21                  TOSHIBA MD04ABA500V        4050
22                 HGST HDS5C4040ALE630        3117
23                 HGST HUS726040ALE610        2598
24                          ST8000DM005        2250
25                      TOSHIBA HDWF180        1798
26   Seagate BarraCuda SSD ZA500CM10002        1564
27                 HGST HUH721010ALE600        1245
28                          ST6000DM001        1113
29                       WDC WD5000BPKT         982
30                      TOSHIBA HDWE160         834
31              Hitachi HDS5C4040ALE630         626
32                          DELLBOSS VD         540
33                          ST8000DM004         273
34  Seagate BarraCuda SSD ZA2000CM10002         265
35                       WDC WD1600AAJS         251
36                          ST6000DM004         180
37                         WDC WD40EFRX          94
38                       ST1000LM024 HN          90
39                 HGST HMS5C4040BLE641          90
40                       WDC WD2500AAJS          88
41                           ST320LT007          85
42                           ST500LM021          82
43                          ST3160316AS          42
44                       WDC WD3200BEKX          29
45              Hitachi HDS724040ALE640          29
46                        ST250LM004 HN          24
47                          ST3160318AS          14
48                       WDC WD3200AAJS           3

In [33]:
# Total drive days across all models (which is also the total rows)
total_drive_days = len(df)
total_drive_days

9577046

### Creating a table that has the number of failures for each model.

### Written in SQL like this:

CREATE TABLE failures AS
    SELECT model, count(*) AS failures
    FROM drive_stats
    WHERE failure = 1
    GROUP BY model;

In [35]:
failures = df[df['failure'] == 1].groupby('model').size().reset_index(name='failures')
failures = failures.sort_values(by='failures', ascending=False).reset_index(drop=True)
failures

model  failures
0          ST12000NM0007       174
1            ST4000DM000       106
2           ST8000NM0055        57
3            ST8000DM002        27
4     TOSHIBA MQ01ABF050        13
5   HGST HMS5C4040BLE640        11
6          ST500LM012 HN         9
7             ST500LM030         8
8   HGST HUH721212ALN604         4
9   HGST HUH728080ALE600         3
10         ST10000NM0086         3
11   TOSHIBA MQ01ABF050M         3
12        WDC WD5000LPVX         2
13        WDC WD5000LPCX         2
14  HGST HMS5C4040ALE640         2
15           ST8000DM004         1
16           ST6000DX000         1
17  HGST HUH721212ALE600         1
18          WDC WD60EFRX         1

In [43]:
# Calculating the total failures
total_failures = (df['failure'] == 1).sum()
print(f'Total failures: {total_failures}')

# Calculating the total unique days
total_unique_days = df['date'].nunique()
print(f'Total unique days: {total_unique_days}')

print(f'Total drive days: {total_drive_days}')

Total failures: 428
Total unique days: 90
Total drive days: 9577046


### So we have 428 drive failures in 9,577,046 drive days of operation.
### The daily failure rate is the drive failures / drive days.
### The annual failure rate would be the daily failure rate * 365 (assuming the rest of the year would have similar results to the first 3 months)

In [44]:
daily_failure_rate = (total_failures / total_drive_days) * 100
annual_failure_rate = daily_failure_rate * 365
print(f'Daily failure rate is {daily_failure_rate}%')
print(f'Annual failure rate is {annual_failure_rate}%')

Daily failure rate is 0.004469018943837171%
Annual failure rate is 1.6311919145005673%


### Creating a table that has the number of drives for each model as of January 31st 2019

### Written in SQL like this:

CREATE TABLE model_count AS
    SELECT model, count(*) AS count
    FROM drive_stats
    WHERE date = '2019-01-31'
    GROUP BY model;

In [51]:
# Convert 'date' column to datetime type (if not already)
df['date'] = pd.to_datetime(df['date'])

# Filter rows for specific date and calculate the number of drives for each model
model_count = df[df['date'] == '2019-01-31'].groupby('model').size().reset_index(name='count')

# Sort the dataframe by 'count' in descending order
model_count = model_count.sort_values(by='count', ascending=False).reset_index(drop=True)
model_count

model  count
0                         ST12000NM0007  32245
1                           ST4000DM000  22854
2                          ST8000NM0055  14383
3                  HGST HMS5C4040BLE640  12880
4                           ST8000DM002   9875
5                  HGST HMS5C4040ALE640   3653
6                  HGST HUH721212ALN604   2475
7                           ST6000DX000   1524
8                   TOSHIBA MG07ACA14TA   1220
9                         ST10000NM0086   1207
10                 HGST HUH728080ALE600   1045
11                        ST500LM012 HN    579
12                   TOSHIBA MQ01ABF050    524
13                         WDC WD60EFRX    382
14                  TOSHIBA MQ01ABF050M    352
15                       WDC WD5000LPVX    250
16                           ST500LM030    153
17                  TOSHIBA MD04ABA400V    144
18                          ST4000DM005     56
19                       WDC WD5000LPCX     55
20                  TOSHIBA MD04ABA500V     45
21                 HGST HDS5C4040ALE630     32
22                 HGST HUS726040ALE610     29
23                          ST8000DM005     25
24                      TOSHIBA HDWF180     20
25   Seagate BarraCuda SSD ZA500CM10002     18
26                          ST6000DM001     13
27                 HGST HUH721010ALE600     13
28                       WDC WD5000BPKT     11
29                      TOSHIBA HDWE160     10
30              Hitachi HDS5C4040ALE630      8
31                          ST8000DM004      3
32  Seagate BarraCuda SSD ZA2000CM10002      3
33                          ST6000DM004      2
34                           ST320LT007      1
35                 HGST HMS5C4040BLE641      1
36                       ST1000LM024 HN      1
37                       WDC WD2500AAJS      1
38                         WDC WD40EFRX      1

### On a specific day, (e.g. 1st Jan, feb, mar) how many hard drives are there for each model

In [47]:
# model_count_jan = df[df['date'] == '2019-01-01'].groupby('model').size().reset_index(name='1st Jan')
# model_count_feb = df[df['date'] == '2019-02-01'].groupby('model').size().reset_index(name='1st Feb')
# model_count_mar = df[df['date'] == '2019-03-01'].groupby('model').size().reset_index(name='1st Mar')
# 
# # Merge the DataFrames on the 'model' column
# model_count = pd.merge(model_count_jan, model_count_feb, on='model', how='outer')
# model_count = pd.merge(model_count, model_count_mar, on='model', how='outer')
# 
# # Fill NaN values with 0 (models that didn't have data for a specific month)
# model_count = model_count.fillna(0)
# 
# # Sort the dataframe by '1st Jan' in descending order
# model_count = model_count.sort_values(by='1st Jan', ascending=False).reset_index(drop=True)
# 
# model_count

model  1st Jan  1st Feb  1st Mar
0                         ST12000NM0007    31147  32246.0  33505.0
1                           ST4000DM000    23236  22853.0  21674.0
2                  HGST HMS5C4040BLE640    14550  12880.0  12774.0
3                          ST8000NM0055    14383  14383.0  14383.0
4                           ST8000DM002     9874   9875.0   9875.0
5                  HGST HMS5C4040ALE640     4678   3654.0   2837.0
6                           ST6000DX000     1524   1524.0   1524.0
7                  HGST HUH721212ALN604     1278   2474.0   3675.0
8                         ST10000NM0086     1210   1207.0   1205.0
9                   TOSHIBA MG07ACA14TA     1204   1160.0   1220.0
10                 HGST HUH728080ALE600     1045   1045.0   1045.0
11                        ST500LM012 HN      604    579.0    544.0
12                   TOSHIBA MQ01ABF050      528    524.0    521.0
13                         WDC WD60EFRX      382    382.0    338.0
14                  TOSHIBA MQ01ABF050M      354    352.0    378.0
15                       WDC WD5000LPVX      272    250.0    235.0
16                  TOSHIBA MD04ABA400V      145    144.0    144.0
17                           ST500LM030      111    152.0    174.0
18                          ST4000DM005       58     56.0     53.0
19                       WDC WD5000LPCX       56     55.0     54.0
20                 HGST HDS5C4040ALE630       50     32.0     28.0
21                  TOSHIBA MD04ABA500V       45     45.0     45.0
22                 HGST HUS726040ALE610       28     29.0     29.0
23                          ST8000DM005       25     25.0     25.0
24                      TOSHIBA HDWF180       20     20.0     20.0
25                       WDC WD1600AAJS       19      0.0      0.0
26              Hitachi HDS5C4040ALE630       14      8.0      2.0
27                          ST6000DM001       13     13.0     13.0
28   Seagate BarraCuda SSD ZA500CM10002       13     18.0     18.0
29                       WDC WD5000BPKT       11     11.0     11.0
30                 HGST HUH721010ALE600       10     13.0     15.0
31                      TOSHIBA HDWE160       10     10.0      9.0
32                          ST8000DM004        4      3.0      3.0
33                          ST3160316AS        3      0.0      0.0
34                          ST6000DM004        2      2.0      2.0
35                         WDC WD40EFRX        2      1.0      1.0
36                          ST3160318AS        1      0.0      0.0
37  Seagate BarraCuda SSD ZA2000CM10002        1      3.0      3.0
38                 HGST HMS5C4040BLE641        1      1.0      1.0
39              Hitachi HDS724040ALE640        1      0.0      0.0
40                       WDC WD2500AAJS        1      1.0      1.0
41                       WDC WD3200AAJS        1      0.0      0.0
42                       WDC WD3200BEKX        1      0.0      0.0
43                       ST1000LM024 HN        1      1.0      1.0
44                        ST250LM004 HN        1      0.0      0.0
45                           ST320LT007        1      1.0      1.0

### Join the tables together and compute the annual failure rate
### drive_years = drive_days / 365
### Annual failure rate = (number of failures / number of drive years) * 100

### Written in SQL like this:

CREATE TABLE failure_rates AS
    SELECT drive_days.model AS model,
           drive_days.drive_days AS drive_days,
           failures.failures AS failures, 
           100.0 * (1.0 * failures) / (drive_days / 365.0) AS annual_failure_rate
    FROM drive_days, failures, model_count
    WHERE drive_days.model = failures.model
      AND model_count.model = failures.model
    ORDER BY model;

In [48]:
drive_days

model  drive_days
0                         ST12000NM0007     2955025
1                           ST4000DM000     1989429
2                          ST8000NM0055     1294451
3                  HGST HMS5C4040BLE640     1172824
4                           ST8000DM002      888741
5                  HGST HMS5C4040ALE640      313383
6                  HGST HUH721212ALN604      259749
7                           ST6000DX000      135832
8                   TOSHIBA MG07ACA14TA      109404
9                         ST10000NM0086      108555
10                 HGST HUH728080ALE600       93598
11                        ST500LM012 HN       50619
12                   TOSHIBA MQ01ABF050       46969
13                  TOSHIBA MQ01ABF050M       32624
14                         WDC WD60EFRX       30523
15                       WDC WD5000LPVX       22015
16                           ST500LM030       14479
17                 HGST HUH721212ALE600       14040
18                  TOSHIBA MD04ABA400V       12662
19                       WDC WD5000LPCX        4920
20                          ST4000DM005        4848
21                  TOSHIBA MD04ABA500V        4050
22                 HGST HDS5C4040ALE630        3117
23                 HGST HUS726040ALE610        2598
24                          ST8000DM005        2250
25                      TOSHIBA HDWF180        1798
26   Seagate BarraCuda SSD ZA500CM10002        1564
27                 HGST HUH721010ALE600        1245
28                          ST6000DM001        1113
29                       WDC WD5000BPKT         982
30                      TOSHIBA HDWE160         834
31              Hitachi HDS5C4040ALE630         626
32                          DELLBOSS VD         540
33                          ST8000DM004         273
34  Seagate BarraCuda SSD ZA2000CM10002         265
35                       WDC WD1600AAJS         251
36                          ST6000DM004         180
37                         WDC WD40EFRX          94
38                       ST1000LM024 HN          90
39                 HGST HMS5C4040BLE641          90
40                       WDC WD2500AAJS          88
41                           ST320LT007          85
42                           ST500LM021          82
43                          ST3160316AS          42
44                       WDC WD3200BEKX          29
45              Hitachi HDS724040ALE640          29
46                        ST250LM004 HN          24
47                          ST3160318AS          14
48                       WDC WD3200AAJS           3

In [49]:
failures

model  failures
0          ST12000NM0007       174
1            ST4000DM000       106
2           ST8000NM0055        57
3            ST8000DM002        27
4     TOSHIBA MQ01ABF050        13
5   HGST HMS5C4040BLE640        11
6          ST500LM012 HN         9
7             ST500LM030         8
8   HGST HUH721212ALN604         4
9   HGST HUH728080ALE600         3
10         ST10000NM0086         3
11   TOSHIBA MQ01ABF050M         3
12        WDC WD5000LPVX         2
13        WDC WD5000LPCX         2
14  HGST HMS5C4040ALE640         2
15           ST8000DM004         1
16           ST6000DX000         1
17  HGST HUH721212ALE600         1
18          WDC WD60EFRX         1

In [52]:
model_count

model  count
0                         ST12000NM0007  32245
1                           ST4000DM000  22854
2                          ST8000NM0055  14383
3                  HGST HMS5C4040BLE640  12880
4                           ST8000DM002   9875
5                  HGST HMS5C4040ALE640   3653
6                  HGST HUH721212ALN604   2475
7                           ST6000DX000   1524
8                   TOSHIBA MG07ACA14TA   1220
9                         ST10000NM0086   1207
10                 HGST HUH728080ALE600   1045
11                        ST500LM012 HN    579
12                   TOSHIBA MQ01ABF050    524
13                         WDC WD60EFRX    382
14                  TOSHIBA MQ01ABF050M    352
15                       WDC WD5000LPVX    250
16                           ST500LM030    153
17                  TOSHIBA MD04ABA400V    144
18                          ST4000DM005     56
19                       WDC WD5000LPCX     55
20                  TOSHIBA MD04ABA500V     45
21                 HGST HDS5C4040ALE630     32
22                 HGST HUS726040ALE610     29
23                          ST8000DM005     25
24                      TOSHIBA HDWF180     20
25   Seagate BarraCuda SSD ZA500CM10002     18
26                          ST6000DM001     13
27                 HGST HUH721010ALE600     13
28                       WDC WD5000BPKT     11
29                      TOSHIBA HDWE160     10
30              Hitachi HDS5C4040ALE630      8
31                          ST8000DM004      3
32  Seagate BarraCuda SSD ZA2000CM10002      3
33                          ST6000DM004      2
34                           ST320LT007      1
35                 HGST HMS5C4040BLE641      1
36                       ST1000LM024 HN      1
37                       WDC WD2500AAJS      1
38                         WDC WD40EFRX      1

In [60]:
# Merge dataframes
# The on='model' argument in the pd.merge function indicates that the merge should be performed 
# based on the model column, and only rows with matching model values in both DataFrames will 
# be included in the result.
merged_df = pd.merge(drive_days, failures, on='model')
merged_df = pd.merge(merged_df, model_count, on='model')
merged_df

model  drive_days  failures  count
0          ST12000NM0007     2955025       174  32245
1            ST4000DM000     1989429       106  22854
2           ST8000NM0055     1294451        57  14383
3   HGST HMS5C4040BLE640     1172824        11  12880
4            ST8000DM002      888741        27   9875
5   HGST HMS5C4040ALE640      313383         2   3653
6   HGST HUH721212ALN604      259749         4   2475
7            ST6000DX000      135832         1   1524
8          ST10000NM0086      108555         3   1207
9   HGST HUH728080ALE600       93598         3   1045
10         ST500LM012 HN       50619         9    579
11    TOSHIBA MQ01ABF050       46969        13    524
12   TOSHIBA MQ01ABF050M       32624         3    352
13          WDC WD60EFRX       30523         1    382
14        WDC WD5000LPVX       22015         2    250
15            ST500LM030       14479         8    153
16        WDC WD5000LPCX        4920         2     55
17           ST8000DM004         273         1      3

In [61]:
# Calculate annual failure rate
failure_rates = merged_df.copy()
failure_rates['drive_years'] = failure_rates['drive_days'] / 365
failure_rates['annual_failure_rate (%)'] = (failure_rates['failures'] / failure_rates['drive_years']) * 100
failure_rates

model  drive_days  failures  count  drive_years  \
0          ST12000NM0007     2955025       174  32245  8095.958904   
1            ST4000DM000     1989429       106  22854  5450.490411   
2           ST8000NM0055     1294451        57  14383  3546.441096   
3   HGST HMS5C4040BLE640     1172824        11  12880  3213.216438   
4            ST8000DM002      888741        27   9875  2434.906849   
5   HGST HMS5C4040ALE640      313383         2   3653   858.583562   
6   HGST HUH721212ALN604      259749         4   2475   711.641096   
7            ST6000DX000      135832         1   1524   372.142466   
8          ST10000NM0086      108555         3   1207   297.410959   
9   HGST HUH728080ALE600       93598         3   1045   256.432877   
10         ST500LM012 HN       50619         9    579   138.682192   
11    TOSHIBA MQ01ABF050       46969        13    524   128.682192   
12   TOSHIBA MQ01ABF050M       32624         3    352    89.380822   
13          WDC WD60EFRX       30523         1    382    83.624658   
14        WDC WD5000LPVX       22015         2    250    60.315068   
15            ST500LM030       14479         8    153    39.668493   
16        WDC WD5000LPCX        4920         2     55    13.479452   
17           ST8000DM004         273         1      3     0.747945   

    annual_failure_rate (%)  
0                  2.149220  
1                  1.944779  
2                  1.607245  
3                  0.342336  
4                  1.108872  
5                  0.232942  
6                  0.562081  
7                  0.268714  
8                  1.008705  
9                  1.169897  
10                 6.489658  
11                10.102408  
12                 3.356425  
13                 1.195820  
14                 3.315921  
15                20.167139  
16                14.837398  
17               133.699634

### Testing out rule 2 on table 6 from https://www.kdd.org/kdd2016/papers/files/adf0849-botezatuA.pdf
### Seems like the information is not true

In [70]:
filtered_df = df[df['smart_197_raw'] >= 2]
filtered_df[['model', 'failure', 'smart_197_raw']].head(10)

model  failure  smart_197_raw
264          ST12000NM0007        0           88.0
440            ST8000DM002        0           24.0
534   HGST HMS5C4040BLE640        0            8.0
719           ST8000NM0055        0           16.0
945            ST4000DM000        0           16.0
1158           ST4000DM000        0           16.0
1446           ST4000DM000        0            8.0
1478  HGST HMS5C4040BLE640        0           56.0
1516           ST4000DM000        0            8.0
2043           ST4000DM000        0           24.0